src/Compilers/Core/Portable/MetadataReference/ModuleMetadata.cs

```diff
@@ -23,10 +23,26 @@ public sealed partial class ModuleMetadata : Metadata
 
         private readonly PEModule _module;
 
-        private ModuleMetadata(PEReader peReader)
+        /// <summary>
+        /// Optional data that should be kept alive as long as this <see cref="ModuleMetadata"/> is alive.  This can be
+        /// useful, for example, if there is backing memory that the metadata depends on that should be kept rooted so it
+        /// doesn't get garbage collected.
+        /// </summary>
+        private readonly IDisposable? _owner;
+
+        /// <summary>
+        /// Whether or not <see cref="_owner"/> should be <see cref="IDisposable.Dispose"/>'d when this object is
+        /// Disposed.  Is controlled by the <c>leaveOpen</c> flag in <see cref="CreateFromStream(Stream, bool)"/>, or
+        /// the <see cref="PEStreamOptions.LeaveOpen"/> flag in <see cref="CreateFromStream(Stream, PEStreamOptions)"/>.
+        /// </summary>
+        private readonly bool _disposeOwner;
+
+        private ModuleMetadata(PEReader peReader, IDisposable? owner = null, bool disposeOwner = false)
             : base(isImageOwner: true, id: MetadataId.CreateNewId())
         {
```

Comment:
```suggestion
        {
            Debug.Assert(!disposeOwner || owner is {});
```

Seems odd if we broke that invariant

Reply:
Very true!  Added.